Fix cudagraph partitioning handling of deterministic fallbacks

eellison opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #169529
* __->__ #168185

The partitioning fallback logic was not querying the full set of unsafe cudagraph nodes, including when deterministic mode enabled.

edit: looks like it exposes other errors in partitioning. potentially that list is over aggressive as well.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @chenyang78 @kadeng @muchulee8 @amjames @chauhang @aakhundov @coconutruben

eellison committed:
Update

[ghstack-poisoned]

eellison committed:
Update

[ghstack-poisoned]

BoyuanFeng commented:
#167613 should not relate to cudagraph partition, or index_put_. W/ or w/o cudagraph partition, the following code would skip cudagraph:

```
skipping cudagraphs due to disabling cudagraphs due to incompatible op aten.index_put_.default Found from   File "/data/users/boyuan/playground/example.py", line 6, in fn
    return x.index_put_([y], z, True)
```


```python
import torch


def fn(x, y, z):
    x = torch.zeros_like(x)
    return x.index_put_([y], z, True)

# Enable deterministic algorithms (as done in --accuracy mode)
torch.use_deterministic_algorithms(True)

fn_c = torch.compile(mode="reduce-overhead")(fn)

def args():
    x = torch.zeros((512, 512), dtype=torch.bool, device="cuda")
    y = torch.arange(512, dtype=torch.int64, device="cuda")
    z = torch.ones((512, 512), dtype=torch.bool, device="cuda")
    return x, y, z

# Run multiple iterations - should not segfault
# The key fix is that index_put_ triggers partitioning when deterministic mode is enabled
for i in range(3):
    out = fn_c(*args())
    eager_out = fn(*args())
    assert torch.allclose(out, eager_out)
```

eellison committed:
Update

[ghstack-poisoned]

eellison committed:
Update on "Fix cudagraph partitioning handling of deterministic fallbacks"


The partitioning fallback logic was not querying the full set of unsafe cudagraph nodes, including when deterministic mode enabled.

edit: looks like it exposes other errors in partitioning. potentially that list is over aggressive as well.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov coconutruben

[ghstack-poisoned]

eellison commented:
replacing with https://github.com/pytorch/pytorch/pull/168185, sorry for churn.
